File: spark/src/main/scala/com/alibaba/graphar/GraphInfo.scala
/** Copyright 2022 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.graphar

import java.io.{File, FileInputStream}
import org.apache.hadoop.fs.{Path, FileSystem}
import org.apache.spark.sql.{SparkSession}
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.Constructor
import scala.beans.BeanProperty

/** Main data type in gar enumeration */
object GarType extends Enumeration{
  type GarType = Value
  /** Boolean type */
  val BOOL = Value(0)
  /** Signed 32-bit integer */
  val INT32 = Value(2)
  /** Signed 64-bit integer */
  val INT64 = Value(3)
  /** 4-byte floating point value */
  val FLOAT = Value(4)
  /** 8-byte floating point value */
  val DOUBLE = Value(5)
  /** UTF8 variable-length string */
  val STRING = Value(6)

  /** Data type in gar to string.
   *
   * @param gar_type gar data type
   * @return string name of gar data type.
   *
   */
  def GarTypeToString(gar_type: GarType.Value): String = gar_type match {
    case GarType.BOOL => "bool"
    case GarType.INT32 => "int32"
    case GarType.INT64 => "int64"
    case GarType.FLOAT => "float"
    case GarType.DOUBLE => "double"
    case GarType.STRING => "string"
    case _ => throw new IllegalArgumentException
  }

  /** String name to data type in gar.
   *
   * @param str the string name of gar data type
   * @return gar data type.
   *
   */
  def StringToGarType(str: String): GarType.Value = str match {
    case "bool" => GarType.BOOL
    case "int32" => GarType.INT32
    case "int64" => GarType.INT64
    case "float" => GarType.FLOAT
    case "double" => GarType.DOUBLE
    case "string" => GarType.STRING
    case _ => throw new IllegalArgumentException
  }
}

/** Type of  file format. */
object FileType extends Enumeration {
  type FileType = Value
  val CSV = Value(0)
  val PARQUET = Value(1)
  val ORC = Value(2)

  /** File type to string.
   *
   * @param file_type  file type in gar
   * @return string name of file type.
   *
   */
  def FileTypeToString(file_type: FileType.Value): String = file_type match {
    case FileType.CSV => "csv"
    case FileType.PARQUET => "parquet"
    case FileType.ORC => "orc"
    case _ => throw new IllegalArgumentException
  }

  /** String to file type.
   *
   * @param str the string name of file type
   * @return file type in gar.
   *
   */
  def StringToFileType(str: String): FileType.Value = str match {
    case "csv" => FileType.CSV
    case "parquet" => FileType.PARQUET
    case "orc" => FileType.ORC
    case _ => throw new IllegalArgumentException
  }

}

/** Adj list type enumeration for adjacency list of graph. */
object AdjListType extends Enumeration {
  type AdjListType = Value
  /** collection of edges by source, but unordered, can represent COO format */
  val unordered_by_source = Value(0)
  /** collection of edges by destination, but unordered, can represent COO format */
  val unordered_by_dest = Value(1)
  /** collection of edges by source, ordered by source, can represent CSR format */
  val ordered_by_source = Value(2)
  /** collection of edges by destination, ordered by destination, can represent CSC format */
  val ordered_by_dest = Value(3)

  /** adjList type in gar to string */
  def AdjListTypeToString(adjList_type: AdjListType.Value): String = adjList_type match {
    case AdjListType.unordered_by_source => "unordered_by_source"
    case AdjListType.unordered_by_dest => "unordered_by_dest"
    case AdjListType.ordered_by_source => "ordered_by_source"
    case AdjListType.ordered_by_dest => "ordered_by_dest"
    case _ => throw new IllegalArgumentException
  }

  /** String to adjList type in gar */
  def StringToAdjListType(str: String): AdjListType.Value = str match {
    case "unordered_by_source" => AdjListType.unordered_by_source
    case "unordered_by_dest" => AdjListType.unordered_by_dest
    case "ordered_by_source" => AdjListType.ordered_by_source
    case "ordered_by_dest" => AdjListType.ordered_by_dest
    case _ => throw new IllegalArgumentException
  }
}

/** The property information of vertex or edge. */
class Property () {
  @BeanProperty var name: String = ""
  @BeanProperty var data_type: String = ""
  @BeanProperty var is_primary: Boolean = false

  /** Get data type in gar of the property */
  def getData_type_in_gar: GarType.Value = {
    GarType.StringToGarType(data_type)
  }
}

/** PropertyGroup is a class to store the property group information. */
class PropertyGroup () {
  @BeanProperty var prefix: String = ""
  @BeanProperty var file_type: String = ""
  @BeanProperty var properties = new java.util.ArrayList[Property]()

  /** Get file type in gar of property group */
  def getFile_type_in_gar: FileType.Value = {
    FileType.StringToFileType(file_type)
  }
}

/** AdjList is a class to store the adj list information of edge. */
class AdjList () {
  @BeanProperty var ordered: Boolean = false
  @BeanProperty var aligned_by: String = "src"
  @BeanProperty var prefix: String = ""
  @BeanProperty var file_type: String = ""
  @BeanProperty var property_groups = new java.util.ArrayList[PropertyGroup]()

  /** Get file type in gar of adj list */
  def getFile_type_in_gar: FileType.Value = {
    FileType.StringToFileType(file_type)
  }

  /** Get adj list type in string*/
  def getAdjList_type: String = {
    var str: String = ""
    if (ordered) {
      str = "ordered_by_"
    } else {
      str = "unordered_by_"
    }
    if (aligned_by == "src") {
      str += "source"
    } else {
      str += "dest"
    }
    return str
  }

  /** Get adj list type in gar */
  def getAdjList_type_in_gar: AdjListType.Value = {
    AdjListType.StringToAdjListType(getAdjList_type)
  }
}

/** GraphInfo is a class to store the graph meta information. */
class GraphInfo() {
  @BeanProperty var name: String = ""
  @BeanProperty var prefix: String = ""
  @BeanProperty var vertices = new java.util.ArrayList[String]()
  @BeanProperty var edges = new java.util.ArrayList[String]()
  @BeanProperty var version: String = ""
}

/** Helper object to load graph info files */
object GraphInfo {
  /** Load a yaml file from path and construct a GraphInfo from it. */
  def loadGraphInfo(graphInfoPath: String, spark: SparkSession): GraphInfo = {
    val path = new Path(graphInfoPath)
    val fs = path.getFileSystem(spark.sparkContext.hadoopConfiguration)
    val input = fs.open(path)
    val yaml = new Yaml(new Constructor(classOf[GraphInfo]))
    val graph_info = yaml.load(input).asInstanceOf[GraphInfo]
    if (graph_info.getPrefix == "") {
      val pos = graphInfoPath.lastIndexOf('/')
      if (pos != -1) {
        val prefix = graphInfoPath.substring(0, pos + 1) // +1 to include the slash
        graph_info.setPrefix(prefix)
      }
    }
    return graph_info
  }
}

File: spark/graphar/src/main/java/com/alibaba/graphar/info/GraphInfo.java
/*
 * Copyright 2022-2023 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.graphar.info;

import com.alibaba.graphar.info.yaml.GraphYamlParser;
import com.alibaba.graphar.util.GeneralParams;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.io.IOException;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.yaml.snakeyaml.LoaderOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;

public class GraphInfo {
    private final String name;
    private final ImmutableList<VertexInfo> vertexInfos;
    private final ImmutableList<EdgeInfo> edgeInfos;
    private final String prefix;
    private final ImmutableMap<String, VertexInfo> vertexLabel2VertexInfo;
    private final ImmutableMap<String, EdgeInfo> edgeConcat2EdgeInfo;
    private final String version;

    public GraphInfo(
            String name,
            List<VertexInfo> vertexInfos,
            List<EdgeInfo> edgeInfos,
            String prefix,
            String version) {
        this.name = name;
        this.vertexInfos =
                vertexInfos instanceof ImmutableList
                        ? (ImmutableList<VertexInfo>) vertexInfos
                        : ImmutableList.copyOf(vertexInfos);
        this.edgeInfos =
                edgeInfos instanceof ImmutableList
                        ? (ImmutableList<EdgeInfo>) edgeInfos
                        : ImmutableList.copyOf(edgeInfos);
        this.prefix = prefix;
        this.version = version;
        this.vertexLabel2VertexInfo =
                vertexInfos.stream()
                        .collect(
                                ImmutableMap.toImmutableMap(
                                        VertexInfo::getLabel, Function.identity()));
        this.edgeConcat2EdgeInfo =
                edgeInfos.stream()
                        .collect(
                                ImmutableMap.toImmutableMap(
                                        EdgeInfo::getConcat, Function.identity()));
    }

    private GraphInfo(GraphYamlParser graphYaml, Configuration conf) throws IOException {
        this(
                graphYaml.getName(),
                vertexFileNames2VertexInfos(graphYaml.getVertices(), conf),
                edgeFileNames2EdgeInfos(graphYaml.getEdges(), conf),
                graphYaml.getPrefix(),
                graphYaml.getVersion());
    }

    private GraphInfo(
            String name,
            ImmutableList<VertexInfo> vertexInfos,
            ImmutableList<EdgeInfo> edgeInfos,
            String prefix,
            ImmutableMap<String, VertexInfo> vertexLabel2VertexInfo,
            ImmutableMap<String, EdgeInfo> edgeConcat2EdgeInfo,
            String version) {
        this.name = name;
        this.vertexInfos = vertexInfos;
        this.edgeInfos = edgeInfos;
        this.prefix = prefix;
        this.vertexLabel2VertexInfo = vertexLabel2VertexInfo;
        this.edgeConcat2EdgeInfo = edgeConcat2EdgeInfo;
        this.version = version;
    }

    public static GraphInfo load(String graphPath) throws IOException {
        return load(graphPath, null);
    }

    public static GraphInfo load(String graphPath, Configuration conf) throws IOException {
        if (graphPath == null) {
            conf = new Configuration();
        }
        Path path = new Path(graphPath);
        FileSystem fileSystem = path.getFileSystem(conf);
        FSDataInputStream inputStream = fileSystem.open(path);
        Yaml graphYamlLoader =
                new Yaml(new Constructor(GraphYamlParser.class, new LoaderOptions()));
        GraphYamlParser graphYaml = graphYamlLoader.load(inputStream);
        return new GraphInfo(graphYaml, conf);
    }

    // TODO(@Thespica): Implement save and dump methods
    //    public void save(String path) {
    //
    //    }
    //
    //    public String dump() {
    //
    //    }

    Optional<GraphInfo> addVertexAsNew(VertexInfo vertexInfo) {
        if (vertexInfo == null || hasVertexInfo(vertexInfo.getLabel())) {
            return Optional.empty();
        }
        ImmutableList<VertexInfo> newVertexInfos =
                ImmutableList.<VertexInfo>builder().addAll(vertexInfos).add(vertexInfo).build();
        ImmutableMap<String, VertexInfo> newVertexLabel2VertexInfo =
                ImmutableMap.<String, VertexInfo>builder()
                        .putAll(vertexLabel2VertexInfo)
                        .put(vertexInfo.getLabel(), vertexInfo)
                        .build();
        return Optional.of(
                new GraphInfo(
                        name,
                        newVertexInfos,
                        edgeInfos,
                        prefix,
                        newVertexLabel2VertexInfo,
                        edgeConcat2EdgeInfo,
                        version));
    }

    Optional<GraphInfo> addEdgeAsNew(EdgeInfo edgeInfo) {
        if (edgeInfo == null
                || hasEdgeInfo(
                        edgeInfo.getSrcLabel(), edgeInfo.getEdgeLabel(), edgeInfo.getDstLabel())) {
            return Optional.empty();
        }
        ImmutableList<EdgeInfo> newEdgeInfos =
                ImmutableList.<EdgeInfo>builder().addAll(edgeInfos).add(edgeInfo).build();
        ImmutableMap<String, EdgeInfo> newEdgeConcat2EdgeInfo =
                ImmutableMap.<String, EdgeInfo>builder()
                        .putAll(edgeConcat2EdgeInfo)
                        .put(edgeInfo.getConcat(), edgeInfo)
                        .build();
        return Optional.of(
                new GraphInfo(
                        name,
                        vertexInfos,
                        newEdgeInfos,
                        prefix,
                        vertexLabel2VertexInfo,
                        newEdgeConcat2EdgeInfo,
                        version));
    }

    public boolean hasVertexInfo(String label) {
        return vertexLabel2VertexInfo.containsKey(label);
    }

    public boolean hasEdgeInfo(String srcLabel, String edgeLabel, String dstLabel) {
        return edgeConcat2EdgeInfo.containsKey(srcLabel + dstLabel + edgeLabel);
    }

    public VertexInfo getVertexInfo(String label) {
        checkVertexExist(label);
        return vertexLabel2VertexInfo.get(label);
    }

    public EdgeInfo getEdgeInfo(String srcLabel, String edgeLabel, String dstLabel) {
        checkEdgeExist(srcLabel, edgeLabel, dstLabel);
        return edgeConcat2EdgeInfo.get(srcLabel + edgeLabel + dstLabel);
    }

    public int getVertexInfoNum() {
        return vertexInfos.size();
    }

    public int getEdgeInfoNum() {
        return edgeInfos.size();
    }

    public String getName() {
        return name;
    }

    public ImmutableList<VertexInfo> getVertexInfos() {
        return vertexInfos;
    }

    public ImmutableList<EdgeInfo> getEdgeInfos() {
        return edgeInfos;
    }

    public String getPrefix() {
        return prefix;
    }

    public String getVersion() {
        return version;
    }

    private static ImmutableList<VertexInfo> vertexFileNames2VertexInfos(
            List<String> vertexFileNames, Configuration conf) throws IOException {
        ImmutableList.Builder<VertexInfo> verticesBuilder = ImmutableList.builder();
        for (String vertexFileName : vertexFileNames) {
            verticesBuilder.add(VertexInfo.load(vertexFileName, conf));
        }
        return verticesBuilder.build();
    }

    private static ImmutableList<EdgeInfo> edgeFileNames2EdgeInfos(
            List<String> edgeFileNames, Configuration conf) throws IOException {
        ImmutableList.Builder<EdgeInfo> edgesBuilder = ImmutableList.builder();
        for (String edgeFileName : edgeFileNames) {
            edgesBuilder.add(EdgeInfo.load(edgeFileName, conf));
        }
        return edgesBuilder.build();
    }

    private void checkVertexExist(String label) {
        if (!hasVertexInfo(label)) {
            throw new IllegalArgumentException(
                    "Vertex label " + label + " not exist in graph " + getName());
        }
    }

    private void checkEdgeExist(String srcLabel, String dstLabel, String edgeLabel) {
        if (!hasEdgeInfo(srcLabel, dstLabel, edgeLabel)) {
            throw new IllegalArgumentException(
                    "Edge label"
                            + srcLabel
                            + GeneralParams.regularSeparator
                            + GeneralParams.regularSeparator
                            + edgeLabel
                            + GeneralParams.regularSeparator
                            + dstLabel
                            + " not exist in graph "
                            + getName());
        }
    }
}
